README.md
openmediavault-smartfancontrol
==============================

SmartFanControl plugin for OpenMediaVault.


debian/changelog
openmediavault-smartfancontrol (5.0) stable; urgency=low

  * Initial release

 -- Helly1206 <hellyrulez@hotmail.com>  Sun, 03 May 2020 09:16:41 +0100


install.sh
#!/bin/bash
NAME="openmediavault-smartfancontrol"
DEBFOLDER="debian"

if [ "$EUID" -ne 0 ]
then 
	echo "Please execute as root ('sudo install.sh' or 'sudo make install')"
	exit
fi

if [ "$1" == "-h" ] || [ "$1" == "-H" ]
then
	echo "Usage:"
	echo "  -h/ -H       : this help file"
	echo "  -d/ -D       : build debian package"
	echo "  -c/ -C       : Cleanup compiled files in install folder"
elif [ "$1" == "-c" ] || [ "$1" == "-C" ]
then
	echo "$NAME Deleting compiled files in install folder"
	rm -f ./*.deb
	rm -rf "$DEBFOLDER"/$NAME
	rm -rf "$DEBFOLDER"/.debhelper
	rm -f "$DEBFOLDER"/files
	rm -f "$DEBFOLDER"/files.new
	rm -f "$DEBFOLDER"/$NAME.*
elif [ "$1" == "-d" ] || [ "$1" == "-D" ]
then
	echo "$NAME build debian package"
	fakeroot debian/rules clean binary
	mv ../*.deb .
else
	echo "$NAME no argument, enter ./install.sh -h for help"
fi


opt/smartfancontrol/omv-smartfancontrol.py
#!/usr/bin/env python3
#
# This file is part of OpenMediaVault.
#
# @license   http://www.gnu.org/licenses/gpl.html GPL Version 3
# @author    Volker Theile <volker.theile@openmediavault.org>
# @copyright Copyright (c) 2009-2020 Volker Theile
#
# OpenMediaVault is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# any later version.
#
# OpenMediaVault is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with OpenMediaVault. If not, see <http://www.gnu.org/licenses/>.

#########################################################
# SCRIPT : omv-smartfancontrol.py                       #
#          OMV database control of smartfancontrol      #
#          settings                                     #
#          I. Helwegen 2020                             #
#########################################################
import sys
import os
import json
import xml.etree.ElementTree as ET
from xml.dom.minidom import parseString
import openmediavault.confdbadm
import openmediavault.config.object
import openmediavault.config.database
from getopt import getopt, GetoptError

ID_SFC = "conf.service.smartfancontrol"
XML_FILE_LOC = "/etc/smartfancontrol.xml"
ENCODING = 'utf-8'
DEFAULTDATA = {'fan': {'mode': 'RPM', 'ONOFFgpio': 27, 'ONOFFinvert': False, 'PWMcalibrated': 30, 'PWMgpio': 18, 'RPMpullup': True, 'PWMfrequency': 10000, 'PWMinvert': False, 'recalibrate': 7, 'RPMgpio': 17, 'RPMppr': 2, 'RPMedge': True, 'RPMfiltersize': 0, 'Frequency': 10, 'Pgain': 0.1, 'Igain': 0.2}, 'temp': {'cpu': True, 'hdd': None, 'ext': None, 'mode': 'MAX', 'Farenheit': False, 'AlarmHigh': 65, 'AlarmCrit': 80, 'AlarmShutdown': True}, 'control': {'mode': 'LINEAR', 'TempOn': 55, 'TempHyst': 5, 'TempStart': 45, 'TempFull': 65, 'LinSteps': 2.5, 'Frequency': 1, 'Pgain': 10.0, 'Igain': 0.1}}
TXTELEMENTS = ["hdd","ext","AlarmShutdown"]   

#########################################################
# Class : Readdb                                        #
#########################################################
class Readdb(openmediavault.confdbadm.ICommand):
    @property
    def description(self):
        return "Reads smartfancontrol settings from database."

    def execute(self, dbid):
        db = openmediavault.config.Database()    
        objs = db.get(dbid, None)
        # Prepare the output.
        if isinstance(objs, list):
            data = [obj.get_dict() for obj in objs]
        else:
            if not isinstance(objs, openmediavault.config.Object):
                return None
            data = objs.get_dict()
        return data
#########################################################

#########################################################
# Class : Writedb                                       #
#########################################################
class Writedb(openmediavault.confdbadm.ICommand):
    @property
    def description(self):
        return "Update smartfancontrol configuration object."

    def execute(self, dbid, data):
        # Create the configuration object.
        obj = openmediavault.config.Object(dbid)
        obj.set_dict(data)
        # Put the configuration object.
        db = openmediavault.config.Database()
        if not db.exists(dbid, None):
            print("ERROR: Database ID doesn't exist")
            exit(2)
        db.set(obj)
        return data
       
        
#########################################################

#########################################################
# Class : XML                                           #
#########################################################
class XML(object):
    def __init__(self):
        self.XMLpath = XML_FILE_LOC
        self.settings = {}
        self.GetXML()
    
    def __del__(self):
        del self.settings
        
    def get(self):
        return self.settings
    
    def set(self, settings):
        self.settings = settings
    
    def GetXML(self):
        try:         
            tree = ET.parse(self.XMLpath)
            root = tree.getroot()
            self.settings = {}
        
            for child in root:
                childdict={}
                childname=child.tag
                for toy in child:
                    if toy.tag in TXTELEMENTS:
                        childdict[toy.tag]=self.gettype(toy.text, True)
                    else:
                        childdict[toy.tag]=self.gettype(toy.text)
                self.settings[childname]=childdict
                
        except Exception as e:
            print("Error parsing xml file")
            print("Check XML file syntax for errors")
            print(e)
            exit(1)
            
    def updateXML(self):
        settings = ET.Element('settings')
        comment = ET.Comment(self.getXMLcomment("settings"))
        settings.append(comment)
        for key, value in self.settings.items():
            child = ET.SubElement(settings, key)
            for key, value in value.items():
                toy = ET.SubElement(child, key)
                toy.text = self.settype(value)
    
        with open(self.XMLpath, "w") as xml_file:
            xml_file.write(self.prettify(settings))
    
    def getXMLcomment(self, tag):
        comment = ""
        with open(self.XMLpath, 'r') as xml_file:
            content = xml_file.read()
            xmltag = "<{}>".format(tag)
            xmlend = "</{}>".format(tag)
            begin = content.find(xmltag)
            end = content.find(xmlend)
            content = content[begin:end]
            cmttag = "<!--"
            cmtend = "-->"
            begin = content.find(cmttag)
            end = content.find(cmtend)
            if (begin > -1) and (end > -1):
                comment = content[begin+len(cmttag):end]
        return comment
    
    def prettify(self, elem):
        """Return a pretty-printed XML string for the Element.
        """
        rough_string = ET.tostring(elem, ENCODING)
        reparsed = parseString(rough_string)
        return reparsed.toprettyxml(indent="\t").replace('<?xml version="1.0" ?>','<?xml version="1.0" encoding="%s"?>' % ENCODING)
    
    def gettype(self, text, txtype = False):
        try:
            retval = int(text)
        except:
            try:
                retval = float(text)
            except:
                if text != None:
                    if txtype:
                        if text.lower() == "false" or text.lower() == "true":
                            retval = str(text).upper()
                        else:
                            retval = str(text).lower()
                    elif text.lower() == "false":
                        retval = False
                    elif text.lower() == "true":
                        retval = True
                    else:
                        retval = str(text)
                else:
                    retval = ""
                        
        return retval
    
    def settype(self, element):
        retval = ""
        if type(element) == bool:
            if element:
                retval = "true"
            else:
                retval = "false"
        elif type(element) == str:
            if element == "":
                retval = None
            else:
                retval = element
        elif element != None:
            retval = str(element)
        
        return retval
#########################################################

def writeXML():
    print("Update settings file from DB (omv changes)")
    readdb = Readdb()
    data = readdb.execute(ID_SFC)
    if data == None:
        print("No data")
    else:
        try:
            data.pop('enable')
        except:
            print("Corrupted data")
            exit(1)
        try:
            data.pop('manual')
        except:
            print("Corrupted data")
            exit(1)
        xml = XML()
        xml.set(data)
        xml.updateXML()
    
def Writeconfig():
    print("Update DB from settings file (manual xml changes)")
    readdb = Readdb()
    rddata = readdb.execute(ID_SFC)
    data = {}

    if rddata == None:
        data['enable'] = False # will be updated from OMV
    else:
        try:
            data['enable'] = rddata['enable']
        except:
            data['enable'] = False
    
    data['manual'] = False # after writing to DB always set manual to false
    
    xml = XML()
    xmldata = xml.get()
    if not xmldata:
        xmldata = DEFAULTDATA
    data.update(xmldata)
    writedb = Writedb()
    writedb.execute(ID_SFC, data)    

def parseopts(argv):
    mode = 0
    print("SmartFanControl fan and temperature control (omv functions)")
    try:
        opts, args = getopt(argv,"hd,",["help","db"])
    except GetoptError:
        print("Enter 'omv-smartfancontrol -h' for help")
        exit(2)
    for opt, arg in opts:
        if opt in ("-h", "--help"):
            print("Usage:")
            print("         ovm-smartfancontrol <args>")
            print("         -h, --help   : this help file")
            print("         -d, --db     : Update DB from settings file (manual xml changes)")
            print("         <no argument>: Update settings file from DB (omv changes)")
            exit()
        elif opt in ("-d", "--db"):
            mode = 1
    return mode
    
#########################################################

######################### MAIN ##########################

if __name__ == "__main__":
    mode = parseopts(sys.argv[1:])
    
    if mode == 1:
        Writeconfig()
    else:
        writeXML()
    exit()

    
    

var/www/openmediavault/js/omv/module/admin/service/smartfancontrol/Settings.js
/**
 * Copyright (C) 2014-2019 OpenMediaVault Plugin Developers
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

// require("js/omv/WorkspaceManager.js")
// require("js/omv/workspace/form/Panel.js")
// require("js/omv/Rpc.js")
// require("js/omv/data/Store.js")
// require("js/omv/data/Model.js")

Ext.define('OMV.module.admin.service.smartfancontrol.settings', {
    extend: 'OMV.workspace.form.Panel',
    requires: [
            'OMV.Rpc'
    ],
    rpcService: 'SmartFanControl',
    rpcGetMethod: 'getSettings',
    rpcSetMethod: 'setSettings',
    
    plugins: [{
        ptype: 'linkedfields',
        correlations: [{
            conditions: [{
                name: 'manual',
                value: false
            }],
            name: ["enable","fan_mode","fan_ONOFFgpio","fan_ONOFFinvert", "fan_PWMcalibrated", "fan_PWMgpio", "fan_PWMfrequency", 
                   "fan_PWMinvert", "fan_recalibrate", "fan_RPMgpio", "fan_RPMpullup", "fan_RPMppr", "fan_RPMedge", "fan_RPMfiltersize", 
                   "fan_Frequency", "fan_Pgain", "fan_Igain", "temp_cpu", "temp_hdd", "temp_ext", "temp_mode", "temp_Farenheit", "temp_AlarmHigh", 
                   "temp_AlarmCrit", "temp_AlarmShutdown", "control_mode", "control_TempOn", "control_TempHyst", "control_FanStart", "control_TempStart", 
                   "control_TempFull", "control_LinSteps", "control_Frequency", "control_Pgain", "control_Igain"],
            properties: ['!readOnly']
        },{
            conditions: [{
                name: 'temp_Farenheit',
                value: false
            }],
            name: ["mon_tempF"],
            properties: ['!show']
        },{
            conditions: [{
                name: 'temp_Farenheit',
                value: true
            }],
            name: ["mon_tempC"],
            properties: ['!show']
        }]
    }],

    hidePagingToolbar: false,
    reloadOnActivate: true,
    
    getFormItems: function() {
        return [{
            xtype: 'fieldset',
            title: _('General settings'),
            defaults: {
                labelSeparator: ''
            },
            items: [{
                xtype: 'checkbox',
                name: 'enable',
                fieldLabel: _('Enable'),
                readOnly: true,
                checked: false
            },{
				xtype: "numberfield",
				name: "mon_tempC",
				fieldLabel: _("Temperature ['C]"),
				allowDecimals: true,
				readOnly: true,
				value: 0
			},{
				xtype: "numberfield",
				name: "mon_tempF",
				fieldLabel: _("Temperature ['F]"),
				allowDecimals: true,
				readOnly: true,
				value: 0
			},{
				xtype: "numberfield",
				name: "mon_rpm",
				fieldLabel: _("Velocity [RPM]"),
				allowDecimals: true,
				readOnly: true,
				value: 0
			},{
				xtype: "numberfield",
				name: "mon_pwm",
				fieldLabel: _("Output [% PWM]"),
				allowDecimals: true,
				readOnly: true,
				value: 0
			},{
                xtype: 'textfield',
                name: 'mon_alarm',
                fieldLabel: _('Alarm'),
                readOnly: true,
                allowBlank: true,
                value: ''
            }]          
        }, {
            xtype: 'fieldset',
            title: _('Fan settings'),
            defaults: {
                labelSeparator: ''
            },
            items: [{
                xtype: 'combo',
                name: 'fan_mode',
                fieldLabel: _('Mode'),
                queryMode: 'local',
                store: [
                    [ 'ONOFF', _('ONOFF') ],
                    [ 'PWM', _('PWM') ],
                    [ 'RPM', _('RPM') ]
                ],
                editable: false,
                readOnly: false,
                value: 'RPM'
            },{
				xtype: "numberfield",
				name: "fan_ONOFFgpio",
				fieldLabel: _("ONOFF gpio"),
				minValue: 0,
				maxValue: 27,
				allowDecimals: false,
				allowBlank: false,
				readOnly: true,
				value: 27
			},{
                xtype: 'checkbox',
                name: 'fan_ONOFFinvert',
                fieldLabel: _('ONOFF invert'),
                readOnly: true,
                checked: false
            },{
				xtype: "numberfield",
				name: "fan_PWMcalibrated",
				fieldLabel: _("PWM calibrated [%]"),
				minValue: 0,
				maxValue: 100,
				allowDecimals: true,
				allowBlank: false,
				readOnly: true,
				value: 30
			},{
				xtype: "numberfield",
				name: "fan_PWMgpio",
				fieldLabel: _("PWM gpio"),
				minValue: 0,
				maxValue: 27,
				allowDecimals: false,
				allowBlank: false,
				readOnly: true,
				value: 18
			},{
				xtype: "numberfield",
				name: "fan_PWMfrequency",
				fieldLabel: _("PWM frequency [Hz]"),
				minValue: 0,
				maxValue: 30000,
				allowDecimals: false,
				allowBlank: false,
				readOnly: true,
				value: 10000
			},{
                xtype: 'checkbox',
                name: 'fan_PWMinvert',
                fieldLabel: _('PWM invert'),
                readOnly: true,
                checked: false
            },{
				xtype: "numberfield",
				name: "fan_recalibrate",
				fieldLabel: _("Recalibrate"),
				minValue: 0,
				maxValue: 365,
				allowDecimals: false,
				allowBlank: false,
				readOnly: true,
				value: 7
			},{
				xtype: "numberfield",
				name: "fan_RPMgpio",
				fieldLabel: _("RPM gpio"),
				minValue: 0,
				maxValue: 27,
				allowDecimals: false,
				allowBlank: false,
				readOnly: true,
				value: 17
			},{
                xtype: 'checkbox',
                name: 'fan_RPMpullup',
                fieldLabel: _('RPM pullup'),
                readOnly: true,
                checked: true
            },{
				xtype: "numberfield",
				name: "fan_RPMppr",
				fieldLabel: _("RPM ppr"),
				minValue: 1,
				maxValue: 4096,
				allowDecimals: false,
				allowBlank: false,
				readOnly: true,
				value: 2
			},{
                xtype: 'checkbox',
                name: 'fan_RPMedge',
                fieldLabel: _('RPM edge'),
                readOnly: true,
                checked: true
            },{
				xtype: "numberfield",
				name: "fan_RPMfiltersize",
				fieldLabel: _("RPM filter size"),
				minValue: 0,
				maxValue: 255,
				allowDecimals: false,
				allowBlank: false,
				readOnly: true,
				value: 0
			},{
				xtype: "numberfield",
				name: "fan_Frequency",
				fieldLabel: _("Frequency [Hz]"),
				minValue: 0,
				maxValue: 100,
				allowDecimals: false,
				allowBlank: false,
				readOnly: true,
				value: 10
			},{
				xtype: "numberfield",
				name: "fan_Pgain",
				fieldLabel: _("P gain"),
				minValue: 0,
				maxValue: 10000,
				allowDecimals: true,
				allowBlank: false,
				readOnly: true,
				value: 0.1
			},{
				xtype: "numberfield",
				name: "fan_Igain",
				fieldLabel: _("I gain"),
				minValue: 0,
				maxValue: 10000,
				allowDecimals: true,
				allowBlank: false,
				readOnly: true,
				value: 0.2
			}]            
        }, {
            xtype: 'fieldset',
            title: _('Temperature settings'),
            defaults: {
                labelSeparator: ''
            },
            items: [{
                xtype: 'checkbox',
                name: 'temp_cpu',
                fieldLabel: _('CPU'),
                readOnly: true,
                checked: true
            },{
                xtype: 'textfield',
                name: 'temp_hdd',
                fieldLabel: _('HDD'),
                readOnly: true,
                allowBlank: true,
                value: ''
            },{
                xtype: 'textfield',
                name: 'temp_ext',
                fieldLabel: _('External'),
                readOnly: true,
                allowBlank: true,
                value: ''
            },{
                xtype: 'combo',
                name: 'temp_mode',
                fieldLabel: _('Mode'),
                queryMode: 'local',
                store: [
                    [ 'MIN', _('Minimum') ],
                    [ 'AVG', _('Average') ],
                    [ 'MAX', _('Maximum') ]
                ],
                editable: false,
                readOnly: true,
                value: 'MAX'
            },{
                xtype: 'checkbox',
                name: 'temp_Farenheit',
                fieldLabel: _('Farenheit'),
                readOnly: true,
                checked: false
            },{
				xtype: "numberfield",
				name: "temp_AlarmHigh",
				fieldLabel: _("High temperature alarm ['C]"),
				minValue: 0,
				maxValue: 150,
				allowDecimals: true,
				allowBlank: false,
				readOnly: true,
				value: 65
			},{
				xtype: "numberfield",
				name: "temp_AlarmCrit",
				fieldLabel: _("Critical temperature alarm ['C]"),
				minValue: 0,
				maxValue: 150,
				allowDecimals: true,
				allowBlank: false,
				readOnly: true,
				value: 80
			},{
                xtype: 'textfield',
                name: 'temp_AlarmShutdown',
                fieldLabel: _('Shutdown on alarm'),
                readOnly: true,
                allowBlank: true,
                value: ''
            }]    
        }, {
            xtype: 'fieldset',
            title: _('Control settings'),
            defaults: {
                labelSeparator: ''
            },
            items: [{
                xtype: 'combo',
                name: 'control_mode',
                fieldLabel: _('Mode'),
                queryMode: 'local',
                store: [
                    [ 'ONOFF', _('ONOFF') ],
                    [ 'LINEAR', _('LINEAR') ],
                    [ 'PI', _('PI') ]
                ],
                editable: false,
                readOnly: true,
                value: 'LINEAR'
            },{
				xtype: "numberfield",
				name: "control_TempOn",
				fieldLabel: _("On temperature ['C]"),
				minValue: 0,
				maxValue: 150,
				allowDecimals: true,
				allowBlank: false,
				readOnly: true,
				value: 55
			},{
				xtype: "numberfield",
				name: "control_TempHyst",
				fieldLabel: _("Hysteresis temperature ['C]"),
				minValue: 0,
				maxValue: 150,
				allowDecimals: true,
				allowBlank: false,
				readOnly: true,
				value: 5
			},{
				xtype: "numberfield",
				name: "control_FanStart",
				fieldLabel: _("Start fan PWM [%]"),
				minValue: 0,
				maxValue: 100,
				allowDecimals: true,
				allowBlank: false,
				readOnly: true,
				value: 20
			},{
				xtype: "numberfield",
				name: "control_TempStart",
				fieldLabel: _("Start temperature ['C]"),
				minValue: 0,
				maxValue: 150,
				allowDecimals: true,
				allowBlank: false,
				readOnly: true,
				value: 45
			},{
				xtype: "numberfield",
				name: "control_TempFull",
				fieldLabel: _("Full temperature ['C]"),
				minValue: 0,
				maxValue: 150,
				allowDecimals: true,
				allowBlank: false,
				readOnly: true,
				value: 65
			},{
				xtype: "numberfield",
				name: "control_LinSteps",
				fieldLabel: _("Linear steps"),
				minValue: 0,
				maxValue: 150,
				allowDecimals: true,
				allowBlank: false,
				readOnly: true,
				value: 2.5
			},{
				xtype: "numberfield",
				name: "control_Frequency",
				fieldLabel: _("Frequency [Hz]"),
				minValue: 0,
				maxValue: 100,
				allowDecimals: false,
				allowBlank: false,
				readOnly: true,
				value: 1
			},{
				xtype: "numberfield",
				name: "control_Pgain",
				fieldLabel: _("P gain"),
				minValue: 0,
				maxValue: 10000,
				allowDecimals: true,
				allowBlank: false,
				readOnly: true,
				value: 10
			},{
				xtype: "numberfield",
				name: "control_Igain",
				fieldLabel: _("I gain"),
				minValue: 0,
				maxValue: 10000,
				allowDecimals: true,
				allowBlank: false,
				readOnly: true,
				value: 0.1
			}]            
        }
    ];
    },
    
    getButtonItems: function() {
        var items = this.callParent(arguments);

        Ext.Array.push(items, [{
            id: this.getId() + '-manual',
            xtype: 'button',
            text: _('Manual tuning'),
            iconCls: 'x-fa fa-pencil',
            handler: Ext.Function.bind(this.onManualButton, this),
            scope: this,
            disabled: false
        },
        {
            id: this.getId() + '-update',
            xtype: 'button',
            text: _('Update manual'),
            iconCls: 'x-fa fa-download',
            handler: Ext.Function.bind(this.onUpdateButton, this),
            scope: this,
            disabled: true
        },
        {
            id: this.getId() + '-refresh',
            xtype: 'button',
            text: _('Refresh'),
            iconCls: 'x-fa fa-refresh',
            handler: Ext.Function.bind(this.onRefreshButton, this),
            scope: this,
            disabled: false
        }]);

        return items;
    },
    
    initComponent: function() {
		var me = this;
		me.callParent(arguments);
		me.on("load", function(c, values) {
			this.setButtonDisabled("manual", values.manual);
            this.setButtonDisabled("update", !values.manual || values.enable);
		}, me);
	},

    onManualButton: function() {
		var me = this;
        this.setButtonDisabled("manual", true);
		// Execute RPC.
		OMV.Rpc.request({
			scope: me,
			callback: function(id, success, response) {
				this.doReload();
			},
			relayErrors: false,
			rpcData: {
				service: "SmartFanControl",
				method: "manual"
			}
		});
	},
    
    onUpdateButton: function() {
		var me = this;
        this.setButtonDisabled("update", true);
		// Execute RPC.
		OMV.Rpc.request({
			scope: me,
			callback: function(id, success, response) {
				this.doReload();
			},
			relayErrors: false,
			rpcData: {
				service: "SmartFanControl",
				method: "update"
			}
		});
	},
    
    onRefreshButton: function() {
        this.doLoad();
	}

});

OMV.WorkspaceManager.registerPanel({
    id: 'smartfancontrol_settings',
    path: '/service/smartfancontrol',
    text: _('Settings'),
    position: 10,
    className: 'OMV.module.admin.service.smartfancontrol.settings'
});


var/www/openmediavault/js/omv/module/admin/service/smartfancontrol/SmartFanControl.js
/**
 * Copyright (C) 2014-2019 OpenMediaVault Plugin Developers
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

// require("js/omv/WorkspaceManager.js")

OMV.WorkspaceManager.registerNode({
    id: 'smartfancontrol',
    path: '/service',
    text: _('SmartFanControl'),
    icon16: 'images/smartfancontrol.png',
    iconSvg: 'images/smartfancontrol.svg'
});


usr/share/openmediavault/confdb/create.d/conf.service.smartfancontrol.sh
#!/bin/sh
#
# @license   http://www.gnu.org/licenses/gpl.html GPL Version 3
# @author    Volker Theile <volker.theile@openmediavault.org>
# @author    OpenMediaVault Plugin Developers <plugins@omv-extras.org>
# @copyright Copyright (c) 2009-2013 Volker Theile
# @copyright Copyright (c) 2013-2019 OpenMediaVault Plugin Developers
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program. If not, see <http://www.gnu.org/licenses/>.

set -e

. /usr/share/openmediavault/scripts/helper-functions

SERVICE_XPATH_NAME="smartfancontrol"
SERVICE_XPATH="/config/services/${SERVICE_XPATH_NAME}"

if ! omv_config_exists "${SERVICE_XPATH}"; then
    omv_config_add_node "/config/services" "${SERVICE_XPATH_NAME}"
    omv_config_add_key "${SERVICE_XPATH}" "enable" "0"
    omv_config_add_key "${SERVICE_XPATH}" "manual" "0"
    omv_config_add_node "${SERVICE_XPATH}" "fan" ""
    omv_config_add_key "${SERVICE_XPATH}/fan" "mode" "RPM"
    omv_config_add_key "${SERVICE_XPATH}/fan" "ONOFFgpio" "27"
    omv_config_add_key "${SERVICE_XPATH}/fan" "ONOFFinvert" "0"
    omv_config_add_key "${SERVICE_XPATH}/fan" "PWMcalibrated" "5.0"
    omv_config_add_key "${SERVICE_XPATH}/fan" "PWMgpio" "18"
    omv_config_add_key "${SERVICE_XPATH}/fan" "PWMfrequency" "10000"
    omv_config_add_key "${SERVICE_XPATH}/fan" "PWMinvert" "0"
    omv_config_add_key "${SERVICE_XPATH}/fan" "recalibrate" "7"
    omv_config_add_key "${SERVICE_XPATH}/fan" "RPMgpio" "17"
    omv_config_add_key "${SERVICE_XPATH}/fan" "RPMpullup" "1"
    omv_config_add_key "${SERVICE_XPATH}/fan" "RPMppr" "2"
    omv_config_add_key "${SERVICE_XPATH}/fan" "RPMedge" "1"
    omv_config_add_key "${SERVICE_XPATH}/fan" "RPMfiltersize" "0"
    omv_config_add_key "${SERVICE_XPATH}/fan" "Frequency" "10"
    omv_config_add_key "${SERVICE_XPATH}/fan" "Pgain" "0.1"
    omv_config_add_key "${SERVICE_XPATH}/fan" "Igain" "0.2"
    omv_config_add_node "${SERVICE_XPATH}" "temp" ""
    omv_config_add_key "${SERVICE_XPATH}/temp" "cpu" "1"
    omv_config_add_key "${SERVICE_XPATH}/temp" "hdd" ""
    omv_config_add_key "${SERVICE_XPATH}/temp" "ext" ""
    omv_config_add_key "${SERVICE_XPATH}/temp" "mode" "MAX"
    omv_config_add_key "${SERVICE_XPATH}/temp" "Farenheit" "0"
    omv_config_add_key "${SERVICE_XPATH}/temp" "AlarmHigh" "65.0"
    omv_config_add_key "${SERVICE_XPATH}/temp" "AlarmCrit" "80.0"
    omv_config_add_key "${SERVICE_XPATH}/temp" "AlarmShutdown" "TRUE"
    omv_config_add_node "${SERVICE_XPATH}" "control" ""
    omv_config_add_key "${SERVICE_XPATH}/control" "mode" "PI"
    omv_config_add_key "${SERVICE_XPATH}/control" "TempOn" "55.0"
    omv_config_add_key "${SERVICE_XPATH}/control" "TempHyst" "5.0"
    omv_config_add_key "${SERVICE_XPATH}/control" "TempStart" "45.0"
    omv_config_add_key "${SERVICE_XPATH}/control" "TempFull" "65.0"
    omv_config_add_key "${SERVICE_XPATH}/control" "LinSteps" "2.5"
    omv_config_add_key "${SERVICE_XPATH}/control" "Frequency" "1"
    omv_config_add_key "${SERVICE_XPATH}/control" "Pgain" "0.1"
    omv_config_add_key "${SERVICE_XPATH}/control" "Igain" "0.1"
    
    /opt/smartfancontrol/omv-smartfancontrol.py -d
fi

exit 0
